File: src/state_parser.py
"""State parser for loop/state.md file."""
import re
from pathlib import Path
from typing import Optional
from dataclasses import dataclass


@dataclass
class LoopState:
    """Represents the state of the autonomous loop."""
    goal: str
    current_task: str
    constraints: str
    iteration_history: str
    status: str


class StateParser:
    """Parses loop/state.md and extracts structured data."""

    def parse_file(self, file_path: str) -> tuple[bool, Optional[LoopState], Optional[str]]:
        """
        Parse state file from given path.
        
        Returns:
            Tuple of (success, state, error_message)
        """
        try:
            path = Path(file_path)
            if not path.exists():
                return False, None, f"State file not found: {file_path}"
            
            content = path.read_text(encoding='utf-8')
            return self.parse_content(content)
        except Exception as e:
            return False, None, f"Failed to read state file: {str(e)}"

    def parse_content(self, content: str) -> tuple[bool, Optional[LoopState], Optional[str]]:
        """
        Parse state content from markdown string.
        
        Returns:
            Tuple of (success, state, error_message)
        """
        try:
            state = LoopState(
                goal=self._extract_section(content, "Goal"),
                current_task=self._extract_section(content, "Current task"),
                constraints=self._extract_section(content, "Constraints"),
                iteration_history=self._extract_section(content, "Iteration History"),
                status=self._extract_section(content, "Status")
            )
            return True, state, None
        except Exception as e:
            return False, None, f"Failed to parse state content: {str(e)}"

    def _extract_section(self, content: str, section_name: str) -> str:
        """Extract a section from markdown content."""
        pattern = rf'^#\s+{re.escape(section_name)}\s*\n(.*?)(?=\n#|\Z)'
        match = re.search(pattern, content, re.MULTILINE | re.DOTALL | re.IGNORECASE)
        return match.group(1).strip() if match else ""

File: src/core/state-parser.ts
import { readFile } from 'fs/promises';
import { LoopState, ParseResult } from '../types/state.js';

/**
 * Parses the loop/state.md file and extracts structured data
 */
export class StateParser {
  /**
   * Parse state file from given path
   */
  async parseFile(filePath: string): Promise<ParseResult> {
    try {
      const content = await readFile(filePath, 'utf-8');
      return this.parseContent(content);
    } catch (error) {
      if ((error as NodeJS.ErrnoException).code === 'ENOENT') {
        return {
          success: false,
          error: `State file not found: ${filePath}`
        };
      }
      return {
        success: false,
        error: `Failed to read state file: ${(error as Error).message}`
      };
    }
  }

  /**
   * Parse state content from markdown string
   */
  parseContent(content: string): ParseResult {
    try {
      const state: LoopState = {
        goal: this.extractSection(content, 'Goal'),
        currentTask: this.extractSection(content, 'Current task'),
        constraints: this.extractSection(content, 'Constraints'),
        iterationHistory: this.extractSection(content, 'Iteration History'),
        status: this.extractSection(content, 'Status')
      };

      return {
        success: true,
        state
      };
    } catch (error) {
      return {
        success: false,
        error: `Failed to parse state content: ${(error as Error).message}`
      };
    }
  }

  /**
   * Extract a section from markdown content
   */
  private extractSection(content: string, sectionName: string): string {
    const regex = new RegExp(`#\\s+${sectionName}\\s*\\n([\\s\\S]*?)(?=\\n#|$)`, 'i');
    const match = content.match(regex);
    return match ? match[1].trim() : '';
  }
}
